### PR TITLE
Add compute_metrics support to GRPOTrainer

### LINKED ISSUE #2959: `compute_metrics` in `GROPTrainer`
### Feature request

Can you please add `compute_metrics` in the `GROPTrainer`?

I can see its available in other, i.e., [DPOTrainer](https://huggingface.co/docs/trl/v0.15.2/dpo_trainer#trl.DPOTrainer.compute_metrics)

### Motivation

1. We can compute accuracy or downstream eval metrics over the evaluation dataset
2. That will help in more fine-grained logging
3. Better metric for `metric_for_best_model`

### Your contribution

I am open to submit a PR

### PR DESCRIPTION
**Pull request by JulesRoussel2001**
# What does this PR do?

This PR adds `compute_metrics` support to `GRPOTrainer`, so users can compute evaluation metrics on held-out data that are independent of the training reward. Closes #2959.

**Motivation.** Reward-based eval metrics (`eval_reward`, etc.) only show whether the optimized signal is increasing, not whether the model is improving at the underlying task. When the reward is a flawed proxy these diverge: in a small GRPO study using this implementation, the training reward saturated at 1.0 while independently measured task accuracy collapsed ([write-up](https://www.lesswrong.com/posts/hBjn9rqgjrktH9LL3/in-open-rlvr-improvement-depends-on-the-instrument-a-small-2)). In such a run, `metric_for_best_model="eval_reward"` selects the worst checkpoint with full confidence. Detecting the divergence requires a task metric computed independently of the reward, which is what `compute_metrics` makes expressible. (Note: #3384 (reward metrics not reaching the best-model check) was a separate bug, fixed in #5811 / v1.5.0. That fix makes `eval_reward` selectable; this PR makes reward-independent metrics definable.)

**Design.** As noted in #2959, a plain passthrough can't work: GRPO's `prediction_step` returns `(loss, None, None)`, so the standard eval loop would silently ignore `compute_metrics`. Instead, completions and the local per-function reward slices are buffered across eval batches in `_generate_and_score_completions`, and `log()` drains the buffers once per eval cycle, builds the `EvalPrediction`, and merges the returned dict as `eval_<name>`, following GRPO's existing metrics-accumulation pattern. `EvalPrediction.predictions` contains the generated completions (list of strings); `EvalPrediction.label_ids` contains the per-function reward tensor of shape `(N, R)`. Distributed handling: `gather_object` runs on all ranks, accumulation is main-process only, and `rewards_per_func` is already globally gathered — no double counting.

**Tests** cover the happy path, the exact `EvalPrediction` input shapes, `compute_metrics=None`, and that reward metrics (`eval_reward`) continue to be logged alongside custom metrics.

**Scope.** How reference answers reach the metric is an open design point (discussion in #2959). RLOO parity is flagged in-thread and deferred to maintainer preference, happy to fold it in here or port it as a follow-up once this lands. A docs example is planned alongside.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the contributor guideline, Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? #2959
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR 
      was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@qgallouedec @lewtun

---

> [!NOTE]
> **Medium Risk**
> Touches GRPO eval logging and distributed gather of completions/rewards. Incorrect buffering could skew eval metrics or logs, but training/loss paths are unchanged.
> 
> **Overview**
> Adds optional `compute_metrics` to `GRPOTrainer` so eval can report task metrics independent of the training reward.
> 
> Because `prediction_step` returns no predictions, completions and per-function rewards are buffered during eval in `_generate_and_score_completions` (gathered on all ranks) and drained in `log()` into an `EvalPrediction` (`predictions` = generated text, `label_ids` = `(N, R)` reward tensor). Returned scalars are logged as `eval_<name>` alongside existing `eval_reward` metrics.
> 
> Tests cover logging, input shapes, `compute_metrics=None`, and that reward metrics still appear.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9a9e51d5bd69f98f6a88805a8b59cf472d49f5a1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by behroozazarkhalili** [comment]
This is a genuinely useful addition and it has been sitting conflicted, so I rebased it on current `main` and resolved the conflict (I have push access via maintainer-can-modify, so the branch is now updated and green on the style gate). Summary of what I checked, in case it helps a maintainer pick it up:

**Conflict:** only one, in the `transformers` import block of `grpo_trainer.py` (`main` added `BitsAndBytesConfig`, this PR added `EvalPrediction`). Resolved by keeping both. Also applied `ruff` to a pre-existing blank-line-whitespace lint in the `log()` flush so the branch passes `ruff check` + `format`.

**Implementation review (looks sound):**
- `compute_metrics` is forwarded to the base `Trainer`, and during eval the trainer buffers completions and the per-function reward tensor, then builds an `EvalPrediction(predictions=<text completions>, label_ids=<(N, R) rewards>)` and calls `compute_metrics` once per eval cycle in `log()`, merging the result as `eval_<name>`.
- Distributed handling is considered: `gather_object(completions)` is a collective run on all ranks, accumulation happens only on the main process, and `rewards_per_func` is already globally gathered, so there is no double counting.
- Three tests cover the happy path, the exact `EvalPrediction` input shapes (text `predictions`, tensor `label_ids`), and `compute_metrics=None` not breaking.

**Why this matters for #3384:** because the returned metrics are logged as `eval_<name>`, they become selectable by `metric_for_best_model`. So a user can define `compute_metrics` that returns a reward or task metric and use it to save the best model, which is exactly the long-standing ask in #3384 (that thread has a lot of +1s). One framing note: this enables the use case via a user-provided function rather than making the already-logged `eval_reward` directly selectable, so if the preferred fix for #3384 is the latter, that would be a small separate change on top. Either way this is the more general mechanism.

@qgallouedec would you be open to reviewing this? Happy to address anything or add a docs example if useful.

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2850,6 +2882,22 @@ def _generate_and_score_completions(
...
+            all_completions = gather_object(completions)
+            if self.accelerator.is_main_process:
+                self._completions_for_compute_metrics.extend(all_completions)
+                # rewards_per_func is already globally gathered (line ~1316); use it directly
+                if self._rewards_for_compute_metrics is None:
+                    self._rewards_for_compute_metrics = rewards_per_func.cpu()
+                else:
+                    self._rewards_for_compute_metrics = torch.cat(
+                        [self._rewards_for_compute_metrics, rewards_per_func.cpu()], dim=0
+                    )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py]: Fixed in 53a0d1b8. The buffers were filled inside an `if self.accelerator.is_main_process:` guard, so on non-main ranks they stayed empty and `log()`'s `and self._completions_for_compute_metrics` check made those ranks skip the `compute_metrics` block, leaving the `eval_<name>` keys out of their logs.

Both inputs are already identical on every rank, so the guard was unnecessary and actually harmful:

```python
all_completions = gather_object(completions)   # full list on every rank
# rewards_per_func is globally gathered by _calculate_rewards -> same matrix on every rank
```

Removing the guard also aligns this block with the file's existing pattern of appending gathered data on all ranks (e.g. `self._logs["completion"].extend(gather_object(completions_text))`), so every rank now accumulates the identical buffer and computes the same custom metrics. cc @qgallouedec

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -2850,6 +2882,24 @@ def _generate_and_score_completions(
...
+            # computes the same custom metrics everywhere and the resulting eval_<name> keys are present in
+            # the logs of every process rather than only the main one.
+            all_completions = gather_object(completions)
+            self._completions_for_compute_metrics.extend(all_completions)
+            if self._rewards_for_compute_metrics is None:
+                self._rewards_for_compute_metrics = rewards_per_func.cpu()
+            else:
+                self._rewards_for_compute_metrics = torch.cat(
+                    [self._rewards_for_compute_metrics, rewards_per_func.cpu()], dim=0
+                )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:2990]: The consistency point is valid: `RLOOTrainer` does duplicate `_generate_and_score_completions` and `log()`, and it currently has no `compute_metrics` path at all, so the two trainers are genuinely out of alignment here.

One scope note before porting it, though. Unlike a bug fix, this adds a **new capability** to a second trainer rather than realigning existing behavior, so it widens this PR beyond "compute_metrics for GRPO." I would rather not expand @JulesRoussel2001's PR unilaterally. @qgallouedec — would you prefer RLOO parity folded into this PR, or tracked as a small follow-up once the GRPO version lands? Happy to do the port either way. cc @kashif

**Comment by JulesRoussel2001** [comment]
Thanks so much @behroozazarkhalili for rebasing this and for the thorough review, I’ve pulled the updated branch, and the conflict resolution looks right to me.

Regarding the RLOO parity point, I’m happy to implement the corresponding changes myself, either as part of this PR or in a follow-up once the GRPO version lands, whichever @qgallouedec prefers. Since RLOO duplicates the relevant generation and logging paths, the implementation and tests should transfer fairly directly.

I've also been using this implementation in a small GRPO study, with a small experiment-specific extension to provide the reference answers needed by the task metrics. It surfaced a concrete reward-hacking case: the optimized reward saturated at 1.0 while task accuracy collapsed, visible only because evaluation metrics were logged independently from the training reward, which is exactly the use case this PR targets. Write-up [here](https://www.lesswrong.com/posts/hBjn9rqgjrktH9LL3/in-open-rlvr-improvement-depends-on-the-instrument-a-small-2). Happy to turn it into a short documentation example if useful.

**Comment by behroozazarkhalili** [comment]
@JulesRoussel2001 sorry for the slow turnaround. Three updates, one of which changes how this PR should be pitched.

**The #3384 framing I gave you is superseded.** #3384 was closed as completed on 2026-07-25, the day after your message. @albertvillanova traced it to a real bug: GRPO's `log()` used to rebuild the log dict (`logs = {**logs, **metrics}`) so trainer-specific eval metrics never reached the best-model check. That was fixed in #5811 (in-place `logs.update(...)`, merged 2026-05-22, shipped in v1.5.0), and I confirmed it is on current main at `grpo_trainer.py:3381`. So `metric_for_best_model="eval_reward"` now works without this PR. In my original review I noted that making `eval_reward` directly selectable would be the alternative route to #3384; that is what happened, so the PR description should stop pointing at a closed issue.

**Your reward-hacking result is the argument that survives, and it is a better one.** Selecting on `eval_reward` only helps when the reward is the thing you actually want to maximize. The case you describe is exactly where that assumption breaks: reward saturating at 1.0 while task accuracy collapses. A run like that would make `metric_for_best_model="eval_reward"` pick the *worst* checkpoint with full confidence. What makes the divergence visible is having a task metric that is computed independently of the training reward, and `compute_metrics` is what makes such a metric expressible at all. I have not reproduced your experiment, so I am describing the mechanism as you reported it, but that is the motivation I would put at the top of the PR description in place of the #3384 reference. Yes to the documentation example, though I would keep it out of this PR (see below).

**CI has never run on this branch.** Both workflow runs on head `52f112a` sit at `conclusion=action_required`, dated 2026-07-24. That is the first-time-contributor approval gate, not a failure, and it needs a maintainer to release it. Worth knowing, because "no green checks" reads as untested when in fact nothing has been attempted. Separately, the branch is now 89 commits behind main, but `git merge-tree` against main reports zero conflicts, so it still merges cleanly and does not need another rebase yet.

**On RLOO: I would do it as a follow-up, not here.** Trainers in this repo are self-contained by design, and the rule that a change to duplicated logic must be propagated is aimed at fixes to logic that already exists in both, not at new features. This PR is 2 files and +196/-0, which is the version most likely to get reviewed; doubling it after it has already waited three weeks works against that. Once it lands, the RLOO version is a near-mechanical mirror, and the docs example fits naturally there or in its own small docs PR.

All of the above is a contributor's read, not a maintainer decision. @qgallouedec the two things that need you: releasing the workflow-approval gate so this gets its first CI run, and a call on RLOO now versus follow-up.

**Comment by JulesRoussel2001** [comment]
Thanks @behroozazarkhalili! All three points are now addressed:

- **Description rewritten:** the #3384 framing is gone, the reward-hacking divergence is now the lead motivation, and the PR is linked to #2959 (the original open request for this feature), which it closes.
- **Design input posted on #2959:** @qgallouedec asked there what EvalPrediction should carry and how reference answers should reach the metric, so I've written up the design rationale for the current predictions/label_ids choice and a proposal for reference passing (extra eval-dataset columns forwarded as kwargs, symmetric with how reward functions receive them), based on how I handled it in practice in the study.
- **RLOO and docs example:** agreed on follow-up — the description now states this explicitly, deferred to maintainer preference.

Good to know about the CI gate and that the branch still merges cleanly, thanks for checking that, as well for flagging the gate alongside the other affected PR.

**Comment by behroozazarkhalili** [comment]
Checked all three, and they hold up.

The description now leads with the reward-hacking divergence and demotes #3384 to a scoped parenthetical (separate bug, fixed in #5811 / v1.5.0), which is the accurate framing. `Closes #2959` is in place, and the Scope paragraph states the RLOO parity and docs-example deferral explicitly rather than leaving it implicit in the thread.

On the reference-answer question you raised on #2959, your proposal is stronger than "symmetric with reward functions" suggests, because it is not merely symmetric, it is the same mechanism. `GRPOTrainer._calculate_rewards` builds its kwargs as:

```python
keys = [key for key in inputs[0] if key not in ["prompt", "completion", "completion_ids"]]
reward_kwargs = {key: [example[key] for example in inputs] for key in keys}
```

Every extra dataset column already reaches reward functions as a kwarg, with no registration and no schema. Forwarding extra eval-dataset columns to `compute_metrics` the same way would not introduce a convention, it would extend the one GRPO already relies on. That seems like the argument most likely to land.

One practical note on the CI state, since it is the thing actually holding you up rather than anything in the code: on the current head SHA the only check that has run is Cursor Bugbot. The full test matrix has never executed on this PR, so its absence of red is not evidence of green. It needs a maintainer to approve workflow runs for a first-time contributor. Worth flagging directly in a comment so it does not sit unnoticed, since nothing else can progress until that happens.

**Comment by behroozazarkhalili** [comment]
This branch is 124 commits behind main, which I should have checked before the comment above.

```
status=diverged  ahead_by=9  behind_by=124
base main commit: 2026-07-24  (DistillationTrainer refactor, #6513)
current main:     2026-08-20  (#6841)
```

`mergeable=true` only means git finds no textual conflict, which reads as healthy on a branch that has not seen a month of main. Correcting a number from my comment above: eleven commits touched `trl/trainer/grpo_trainer.py` in that window, not five. Five was the count of filenames matching `grpo_trainer.py` in the aggregate diff, which swept in test files and the experimental variants, so it understated the real churn.

The one most likely to reach you is #6534, which merged about four hours after your base commit and adds a hard failure inside `_calculate_rewards`:

```python
if len(output_reward_func) != len(prompts):
    raise ValueError(
        f"The reward function '{reward_func_name}' returned {len(output_reward_func)} rewards, but "
        f"{len(prompts)} were expected (one per prompt-completion pair). ..."
    )
```

That is the same method your per-function reward slices are buffered from, and it now raises where it previously accepted a length mismatch. #6648 (+8/-12) and #6024 (+3/-0) also moved loss and metric normalization in that file.

So the order I suggested is backwards. Merge main first, then ask for the workflow approval, because approving now spends the maintainer action on a July base and a green result would say nothing about current main.

The snippet I quoted does survive the gap: the `_calculate_rewards` kwargs construction is identical at line 1621 on your base and line 1630 on main, so the argument for reference answers as forwarded columns holds either way.

**Comment by JulesRoussel2001** [comment]
Branch is now updated to current main (merged through #6838; previously based on July 24). The GRPO touchpoints from the interim commits (#6534's reward-shape validation in _calculate_rewards, and the loss/metric-normalization changes from #6648 and #6024) required no adjustments: all four compute_metrics tests pass locally on the merged branch.

@qgallouedec with the branch on current main, could you approve the workflow runs when you get a chance? CI has never executed on this PR (first-time-contributor gate), so this would be its first full run.

Thanks @behroozazarkhalili for catching the drift before the approval was spent on a stale base. And for the _calculate_rewards observation, I've added a note on #2959, since it makes the reference-answer proposal a direct extension of the existing kwargs mechanism rather than a new convention.

**Comment by behroozazarkhalili** [comment]
@JulesRoussel2001 after talking with @qgallouedec I reopened this work as #7144 from my fork so it can be reviewed. Your commits are cherry-picked there with your authorship intact and the PR body credits you; if you would rather carry it yourself, say so there and I will close mine.

**Comment by JulesRoussel2001** [comment]
@behroozazarkhalili  Thanks again for stepping in on this. Out of curiosity, since the PR was closed without comment: did Quentin mention anything about the reason when you talked, or anything I should do differently in future contributions? He'd engaged with the design in #2959 in August, so the closure surprised me.
